### ISSUE TITLE
Syntax for adding other HTTP methods to URLFETCH

### ISSUE THREAD

**Issue by randomboolean** [senx/warp10-platform#880]
Currently the URLFETCH fonction only support GET method. The signatures are:
```
url:STRING headers:MAP<STRING> URLFETCH
urls:LIST<STRING> headers:MAP<STRING> URLFETCH
```
with _headers_ being optional.

To support additional methods, I suggest:

```
request:MAP<STRING> headers:MAP<STRING> URLFETCH
requests:LIST<MAP<STRING>> headers:MAP<STRING> URLFETCH
```

where _request_ shall have the fields _'url'_, _'method'_ and optionally _'message'_, and _headers_ are still being optional.

Is that signature good ?

**Comment by ftence** [comment]
It looks good, even if I think the keys of the map could have a different name (_url_, _verb_ and _body_ or _payload_).

Too bad I added the header signature without thinking about that kind of signature before because it would definitively make sense to have also a _header_ key in the map you want to define. As it is a fresh addition (2.7.0) maybe should should consider breaking the header signature to allow 
```
request:MAP<STRING> URLFETCH
requests:LIST<MAP<STRING>> URLFETCH
```
with `requests` containing a _header_ key.

**Comment by randomboolean** [comment]
Not sure about _verb_, I think _method_ is more explicit. I agree with _body_ or _payload_ instead of _message_ though. I think _body_ is better.

Putting a MAP inside a MAP makes for a bit clunky codes.
Maybe, we can add only the signature `request:MAP<STRING> headers:MAP<STRING> URLFETCH`, and not the second one, since anyway, the second signature is not that much of a gain compared to using FOREACH or LMAP, and it doesn't always make sense to use the same headers for all requests.

**Comment by ftence** [comment]
Agreed on url/method/body keys.

I'm not convinced that maps of maps are a problem but breaking the 2.7.0 is more of a problem, I think @hbs is best to arbitrate on that. If we want to keep backward-compatibility I think both  
```
request:MAP<STRING> headers:MAP<STRING> URLFETCH
requests:LIST<MAP<STRING>> headers:MAP<STRING> URLFETCH
```
are nice to have, even if, I do agree, it can be done with `FOREACH` or `LMAP`.

**Comment by randomboolean** [comment]
Also,there might be an issue for the signature `request:MAP<STRING> headers:MAP<STRING> URLFETCH`. Since _headers_ is optional, URLFETCH would have to check if the first map it retrieves from the stack is a _headers_ or _request_. I don't see a way to do this without limiting the names of the fields that are allowed in _headers_.

**Comment by trepca** [comment]
I intend to run many URLFETCH at once (it's a data platform and users can run their scripts as they please) which can quickly hog all memory so a streaming option could be really useful. Otherwise, I need to limit URLFETCH maxsize which is not ideal.

Maybe something like `requests:LIST<MAP<STRING>> headers:MAP<STRING> chunk_size:LONG URLFETCH`. `chunk_size` can be -1 by default but if set, it returns chunks of content with that size as an iterator.

Another way could be `requests:LIST<MAP<STRING>> headers:MAP<STRING> start_byte:LONG end_byte:LONG URLFETCH end:BOOL content:STRING`. When `start_byte` and `end_byte` are defined URLFETCH reads and drops all data before start and reads until `end_byte`. If the end of the stream is reached, `end` is set to True.   

I expect it to be slower but scales much better.

**Comment by ftence** [comment]
Maybe an idea similar to https://github.com/senx/warp10-platform/pull/855 can be used with a byte delimiter (or a chunk size) and a macro to be run for each of these chunks.

The other way around, for POST, a macro could create chunks to be sent, but that may be overengineering.

**Comment by randomboolean** [comment]
This suggestion was added in PR #891

**Comment by hbs** [comment]
Solved by PR #891
